controllers/message-utils.js

```diff
@@ -1,20 +1,97 @@
-var db = require('../db');
-
-var updateCouchDB = function(record_id, body, callback) {
-  db.medic.updateWithHandler(
-    'medic',
-    'update_message_task',
-    record_id,
-    body,
-    callback
-  );
-};
+const _ = require('underscore'),
+      db = require('../db');
 
-var getTaskMessages = function(options, callback) {
+const getTaskMessages = function(options, callback) {
   db.medic.view('medic', 'tasks_messages', options, callback);
 };
 
+// copied from kujua-utils
+// TODO: get rid of this copy once Milan's refactor is complete
+const setTaskState = function(task, state, details) {
+  task.state = state;
+  task.state_details = details;
+  task.state_history = task.state_history || [];
+  task.state_history.push({
+    state: state,
+    state_details: details,
+    timestamp: new Date().toISOString()
+  });
+};
+
+const getTaskForMessage = function(uuid, doc) {
+  const getTaskFromMessage = (tasks, uuid) => tasks.find(task => {
+    if (task.messages) {
+      return task.messages.find(message => uuid === message.uuid);
+    }
+  });
+
+  return getTaskFromMessage(doc.tasks || [], uuid) ||
+         getTaskFromMessage(doc.scheduled_tasks || [], uuid);
+};
+
+
+const getTaskAndDocForMessage = function (messageId, docs) {
+  for (const doc of docs) {
+    const task = getTaskForMessage(messageId, doc);
+    if (task) {
+      return [task, doc._id];
+    }
+  }
+};
+
+/*
+ * Applies (in-place) state changes to a given collection of docs.
+ *
+ * Also returns a map of docId -> taskStateChanges
+*/
+const applyTaskStateChangesToDocs = (taskStateChanges, docs) => {
+  const taskStateChangesByDocId = {};
+  const fillTaskStateChangeByDocId = (taskStateChange, docId) => {
+    if (!taskStateChangesByDocId[docId]) {
+      taskStateChangesByDocId[docId] = [];
+    }
+    taskStateChangesByDocId[docId].push(taskStateChange);
+  };
+
+  taskStateChanges.forEach(taskStateChange => {
+    if (!taskStateChange.messageId) {
+      throw Error('Message id required');
+    }
+
+    const [task, docId] = getTaskAndDocForMessage(taskStateChange.messageId, docs);
+
+    if (!task) {
+      throw Error(`Message not found: ${taskStateChange.messageId}`);
+    }
+
+    fillTaskStateChangeByDocId(taskStateChange, docId);
+    setTaskState(task, taskStateChange.state, taskStateChange.details);
+  });
+
+  return taskStateChangesByDocId;
+};
+
 module.exports = {
+  /*
+   * Gets a message given an id
+   */
+  getMessage: function(id, callback) {
+    if (!id) {
+      return callback({ code: 500, message: 'Missing "id" parameter.' });
+    }
+    getTaskMessages({ key: id }, function(err, data) {
+      if (err) {
+        return callback(err);
+      }
+      if (data.rows.length === 0) {
+        return callback({ code: 404, message: 'Not Found' });
+      }
+      callback(null, data.rows[0].value);
+    });
+  },
+  /*
+   * Gets all messages based on passed state
```

Comment:
I think the word `passed` is unnecessary here, and every time I read it I parse it as _past_, not _passed_.  Also, neither `state` nor `states` options are mandatory, so this comment may be misleading.